logo.js

```diff
@@ -924,84 +924,38 @@ function LogoInterpreter(turtle, stream, savehook)
     self.forceBye = true;
   };
 
-  self.run = function(string, options) {
-    if (self._pendingPromise) {
-      var runner = self.run.bind(self, string, options);
-      return self._queueCall(runner);
-    }
-    options = Object(options);
-    // Parse it
-    var atoms = parse(string);
-
-    if (self.turtle) { self.turtle.begin(); }
-    // And execute it!
-    var promise = self.execute(atoms, options).then(function (result) {
-      if (self.turtle) {
-        self.turtle.end();
-      }
-      return result;
-    }, function (err) {
-      if (self.turtle) {
-        self.turtle.end();
-      }
-      if (err && (err.special == "bye" || err.special == "stop")) {
-        return undefined;
-      }
-      throw err;
-    });
-    this._setPendingPromise(promise);
-    self._pendingPromise = promise;
-    return promise;
-  };
+  function noop() {}
+  var lastRun = Promise.resolve();
 
-  //----------------------------------------------------------------------
-  // Helpers to ensure calls to run() are serialized:
-  //----------------------------------------------------------------------
+  self.run = function(string, options) {
+    var promise = lastRun.then(function() {
```

Comment:
My code replaces lastRun every time, so they will always run in order. No need to clear it either, since .then() will execute immediately (well, next microtask...) if the promise has already resolved.